### PR TITLE
Add queries stats to _cluster/stats API

### PR DESCRIPTION
**Pull request by mayya-sharipova**
- Query types are counted by a coordinating node
- Query types are used in the original form as submitted by a user.
- Query types are deduplicated per search request (e.g. if a search request has 1 bool query and 5 term queries, this is counted as 1 bool query and 1 term query).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 1]
```diff
@@ -147,7 +150,8 @@ public TransportSearchAction(
         ActionFilters actionFilters,
         IndexNameExpressionResolver indexNameExpressionResolver,
         NamedWriteableRegistry namedWriteableRegistry,
-        ExecutorSelector executorSelector
+        ExecutorSelector executorSelector,
+        QueriesUsageService queryUsageService
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:154]: When I look at this in IntelliJ, I see a warning "Class 'QueriesUsageService' is not exported from module 'org.elasticsearch.server'".

**Review by jtibshirani — COMMENTED** [review]
This looks good to me overall!

I had one idea for simplification: do we need a separate `TrackingQueryRewriteContext` and associated methods? Or can we just add usage tracking to `CoordinatorRewriteContext`?

**Review by javanna — CHANGES_REQUESTED** [review]
Sorry about the wait, I left a couple of comments around docs and API output. I had a similar thought as Julie around using coordinator rewrite context, but the problem there is that the context is only used conditionally while queries we always want to track. 

The fact that we track queries as part of rewrite does not feel super natural to me, and the need for additional methods in IndicesService and SearchService is a bit odd. I've been thinking that tracking queries while parsing them would feel more natural: a query comes in from the REST layer, we are parsing it and at the same time tracking what we are parsing. That may look like it requires broad changes to the infra around parsing queries, as we don't carry around any shared parsing context, but there may be another way, I am experimenting with this and I will post an idea around this.

**Comment by javanna** [comment]
Like I said above, I've been looking into tracking queries while parsing them. I see the following options:

1) add a context argument to all the *QueryBuilder#fromXContent methods . This is the ideal way to share context, where we could add the query tracking logic similar to what we are doing in this PR with the rewrite context. This change would touch quite some classes but is mostly mechanical, the main disadvantage I see is that it is a breaking change for plugin developers, as the context argument would become required for queries that are provided by plugins. We can still go ahead and make this change if we think it make sense, but we should look at alternatives too.

2) wrap the xcontent parser and specialize the named object parsing, which is where we can track queries that get parsed: I opened a draft PR to show what I mean: https://github.com/elastic/elasticsearch/pull/90425 . 

While on one hand this is additional effort, I think we should try to add the query tracking where it fits best. One consequence of doing the tracking at parsing is also around counting what comes at REST, so not counting CCS calls twice, and counting incoming queries as they are, before any rewrite.

What do you think @jtibshirani @mayya-sharipova ?

**Comment by jtibshirani** [comment]
I thought about it more, and noticed a surprising behavior when using rewrite to track. If the query rewrites on the coordinating node, then we record both the original and rewritten forms. This is because we call `AbstractQueryBuilder#rewrite` again and again until the query doesn't change. I tried the current code with a [terms lookup query](https://www.elastic.co/guide/en/elasticsearch/reference/8.4/query-dsl-terms-query.html#query-dsl-terms-lookup) that doesn't match any documents. In the stats you see both the original query and the rewritten form: `{terms=1, match_none=1}`.

To me using parsing makes sense and feels low overhead. I don't have an idea for the cleanest way to do it though. 

Another option would be to add a method `QueryBuilder#visit` that lets you visit the whole query tree and collect types. We have the same thing for Lucene queries `Query#visit`, and use it to enforce the total clauses limit. To me this seems like the clearest design, and might not be too much work because we could have a default implementation on `AbstractQueryBuilder` (?)

**Review by DaveCTurner — COMMENTED** [review]
I wonder if this'd make more sense in the node stats API (perhaps also adding these summary  stats to the cluster stats API). I can imagine cases where it would be useful to see this broken down by node. Does it need to live alongside node stats like this? If not, could we move the `QueriesUsageService` into the `NodeService`, and move `QueryStats` into `NodeStats`.

**Comment by javanna** [comment]
> I wonder if this'd make more sense in the node stats API (perhaps also adding these summary stats to the cluster stats API). I can imagine cases where it would be useful to see this broken down by node.

Hey @DaveCTurner we have discussed this in length with the team and decided that there is no value in having these stats part of nodes stats, because what we want to track is incoming queries at the cluster level, and not which queries are received by which coordinating node. It could be different if we were tracking queries on the data nodes, which we did in a previous iteration but we decided to move to the coordinating node based on product requirements.

**Comment by javanna** [comment]
@jtibshirani thanks for digging, I do see how counting on rewrite can lead to surprising behaviors. 

The `visit` idea is something I considered too. I am not sure that we should introduce a separate internal API to visit queries, given that we can do what we need at parse time: in practice we would parse incoming queries, and immediately after traverse the tree of what we just parsed to track queries usage? Could the visit mechanism be used for other purposes in the future? It could be useful if we wanted to analyze the query tree after rewrite, but we could then introduce it later?

@romseygeek you may have thoughts on this.

**Comment by romseygeek** [comment]
I've thought in the past that a `visit` API might be helpful, but actually I think the parser-wrapping PR that Luca linked here is a nicer way to implement it.  I'd be interested to see how the depth validation in the PR combines with query type tracking - possibly we can abstract it further and have something like a query visiting parser that takes a collection of callbacks to be triggered each time we move to a subquery.

**Comment by mayya-sharipova** [comment]
> I thought about it more, and noticed a surprising behavior when using rewrite to track. If the query rewrites on the coordinating node, then we record both the original and rewritten forms. This is because we call AbstractQueryBuilder#rewrite again and again until the query doesn't change. I tried the current code with a [terms lookup query](https://www.elastic.co/guide/en/elasticsearch/reference/8.4/query-dsl-terms-query.html#query-dsl-terms-lookup) that doesn't match any documents. In the stats you see both the original query and the rewritten form: {terms=1, match_none=1}.

@jtibshirani That's a very good point. I was under impression that all queries we need to have a shard context for a query to be rewritten; while on a coordinating node they never get rewritten. But as an example with `terms` query demonstrated this is not true. 

I  agree this is a major downside of counting queries on rewrite, and  we should explore an alternative approach.

**Comment by mayya-sharipova** [comment]
I've synchronized with @javanna offline, and we thought the best way forward would be to proceed with his parsing approach https://github.com/elastic/elasticsearch/pull/90425, but there is a downside with it is that not all queries are going though XContent parsing:
- queries submitted by internal applications (EQL, ML, Security) may not go through parsing. Is it important to count them? We are inclined to think that it is not important, as a goal of this stats would be to count user submitted queries.
- in tests we usually don't use parsing. This is not a blocker, as we can design tests with parsing.

**Pull request closed without merging**